Playwright tests

sinisaos opened the pull request:
Related to #226. `playwright install` needs to be run to install the required browsers. The tests are used by running `./scripts/dev-server` in one terminal and in second terminal run tests with `./scripts/run-e2e-test.sh`.

dantownsend commented on e2e/test_playwright.py:
```diff
@@ -0,0 +1,337 @@
+    context = browser.new_context()
+    # Open new page
+    page = context.new_page()
+    # Go to http://localhost:3000/#/login?nextURL=%2F
+    page.goto("http://localhost:3000/#/login?nextURL=%2F")
```
Out of interest, did you generate these using Playwright Codegen?

If so, how did you find it?

sinisaos replied:
Sorry for the late reply. Yes, tests are generated from `playwright codegen localhost:3000`. After that, two windows were opened. One where you can interact with the web page and one where the code is generated. I had to change each test a bit later, but the tests are 80% generated with Playwright and I think it's a great feature that is very useful.

dantownsend commented on scripts/run-tests.sh:
```diff
@@ -9,4 +9,4 @@ mkdir -p piccolo_admin/dist/css
 touch piccolo_admin/dist/index.html
 
-python -m pytest --cov=piccolo_admin --cov-report xml --cov-report html --cov-fail-under 85 -s $@
+python -m pytest tests --cov=piccolo_admin --cov-report xml --cov-report html --cov-fail-under 85 -s $@
```
Hmm, this is a tricky one.

 I guess you added this so Pytest wouldn't also run the Playwright tests.

The only downside is we can no longer specify a specific test to run (passed in via $@).

```
./run-tests.sh tests/test_foo.py
```

sinisaos replied:
Yes, you're right. I had to add that because `pytest` runs both tests at once.

dantownsend replied:
Maybe instead of specifying 'tests', we can ignore the integration folder instead.

Something like `pytest --ignore=e2e`

dantownsend commented:
Things to do:

- [ ] We need to update `contributing.rst` so it mentions Playwright rather than Cypress
- [ ] Remove Cypress
- [ ] Add GitHub Actions

sinisaos commented:
@dantownsend Setting up Playwright in GH actions was not an easy task. The main problem was running the dev server and the tests in parallel at once. Luckily I found this [snippet from Simon Willison](https://til.simonwillison.net/pytest/playwright-pytest) which I modified a bit and added a database reset. This allows us to start the dev server and tests in parallel with `./scripts/run-e2e-test.sh`. The second step was to set up GH actions. For this I found a great tool to run gh actions locally called [act](https://github.com/nektos/act) and it is very useful.

I have one more question. Do we even use Storybook or can I remove it as well to reduce npm dependencies?

dantownsend commented on e2e/test_playwright.py:
```diff
@@ -0,0 +1,370 @@
+            time.sleep(1)
+            retries -= 1
+
+    Director.delete(force=True).run_sync()
+    Director.insert(*[Director(**d) for d in DIRECTORS]).run_sync()
```
We might not need to do this anymore - it looks like each test starts the server again, so it will already have fresh data each time.

dantownsend replied:
Let's keep Storybook for now - otherwise this PR gets quite big.

Maybe we drop it in the future. I'm torn on Storybook. It can be useful during development sometimes, and encourages us to write nicely isolated components. But it's a pain to configure using Webpack.

dantownsend commented on scripts/run-e2e-test.sh:
```diff
@@ -0,0 +1,6 @@
+
+export PICCOLO_CONF="e2e.piccolo_conf"
+
+python -m pytest e2e 
```
One of the main differences between Playwright and Cypress is Playwright doesn't emit images / videos by default. We need to pass in the `--screenshot=on`, and `--video=on` flags.

https://playwright.dev/python/docs/test-runners#cli-arguments

It would be nice to upload these as build artefacts (not sure if screenshots/videos/both, and whether just on failure). It will help us see why tests fail.

The Cypress GitHub action had some code we can reuse:

```
      - name: "Upload Cypress Screenshots and Videos"
        uses: actions/upload-artifact@v3
        with:
          name: cypress-artifacts
          path: |
            /home/runner/work/piccolo_admin/piccolo_admin/admin_ui/cypress/videos/
            /home/runner/work/piccolo_admin/piccolo_admin/admin_ui/cypress/screenshots/
        if: always()
```

sinisaos replied:
Ok, I'll try adding that and see how it works. Now [act](https://github.com/nektos/act) comes very handy because I can try everything locally, not by pushing commits to Github.

sinisaos replied:
I tried `--screenshot=only-on-failure` and `--screenshot=on` . I made `screenshots` folder inside `e2e` and use
```yaml
- name: "Upload Playwright Screenshots"
  uses: actions/upload-artifact@v3
  with:
    name: playwright-artifacts
    path: |
      /home/runner/work/piccolo_admin/piccolo_admin/e2e/screenshots/
  if: always()
```
and 
```yaml
- uses: actions/upload-artifact@v2
  if: always()
  with:
    name: screenshots
    path: screenshots/
    retention-days: 5
```
but result is always same. I get this warning `::warning::No files were found with the provided path: /home/runner/work/piccolo_admin/piccolo_admin/e2e/screenshots/. No artifacts will be uploaded.` and `screenshots` folder is empty.
In the end I put `--screenshot=only-on-failure`. Do you have any idea how to save screenshots?

dantownsend replied:
I think the folder name might be `test-results' instead of `screenshots`.

For some other Playwright tests I have, the path looks like this:

```
test-results/integration-test-signup-py-test-signup-chromium/test-failed-1.png
```

I think it' possible to customise the path using the `--output` flag.

sinisaos replied:
Sorry but I was unable to set the `test-results`. I don't know where the problem is and I always get `::warning::No files were found with the provided path` on `act` tool. I tried to put `test-results` in `e2e` and in `root`. I tried `--output` flag but nothing works. GH actions says that artifact is successfully uploaded but files size is 0. If you can fix it somehow, that would be great. Apart from that (all tests and GH actions) everything else works fine.

sinisaos commented:
> I did a little bit of work on Storybook fairly recently because it was broken. Honestly, I barely use it, and getting webpack to work properly with it is a pain. I'm OK with keeping it for now, but might drop it in the future.

OK, I won't delete Storybook.

sinisaos commented:
@dantownsend I ended up using videos and the video artifacts were uploaded successfully.

dantownsend commented:
@sinisaos That's great - videos are better anyway.

sinisaos commented:
@dantownsend If you have nothing to add to this PR feel free to merge this.

dantownsend commented:
Cool, it looks good to me.
